api/server/router/container/backend.go

```diff
@@ -43,7 +43,7 @@ type stateBackend interface {
 	ContainerStop(name string, seconds int) error
 	ContainerUnpause(name string) error
 	ContainerUpdate(name string, hostConfig *container.HostConfig) ([]string, error)
-	ContainerWait(name string, timeout time.Duration) (int, error)
+	ContainerWait(name string, timeout time.Duration, waitForRemoval bool) (int, error)
```

Comment:
I think the detach event would be helpful regardless, if you're keen on making the PR, please do so.

For the other matter, during detach, we should just block, the same way we do today, until we get the `die` event. If the container get restarted we should still get a `die` event, and the exit code form this one would be correct the correct one to report (and match today behavior). The timeout wouldn't be needed, since we wouldn't call `ContainerWait` at all anymore.

In the end, I'll defer to @cpuguy83 since he seemed to have a strong opposition to the change in wait :)